File: MazePicSolver/Helpers/ImageHelper.vb
﻿Imports System.Diagnostics.CodeAnalysis
Imports System.Drawing
Imports System.Drawing.Imaging
Imports MazePicSolver.Extensions
Imports MazePicSolver.Map

Namespace Helpers

    Module ImageHelper
        Public Function LoadMapDotsFromImage(imagePath As String) As MapDot()()
            Dim dots As MapDot()()
            Using image = New Bitmap(imagePath)
                dots = Enumerable.Range(0, image.Size.Width).Select(Function(t) New MapDot(image.Size.Height - 1) {}).ToArray()
                ' this style looks a lot nicer in c# and I know the image will box of the heap but minor slow down and small price to pay for dryness
                dots.IterateThroughMap(Sub(x, y, map)
                                           If image.GetPixel(x, y).R = 0 AndAlso image.GetPixel(x, y).G = 0 AndAlso image.GetPixel(x, y).B = 0 Then
                                               map(x)(y) = New MapDot(True, New Point(x, y))
                                           ElseIf image.GetPixel(x, y).G > 0 AndAlso image.GetPixel(x, y).R < 255 AndAlso image.GetPixel(x, y).G <= 255 Then
                                               map(x)(y) = New MapDot(False, New Point(x, y), True)
                                           ElseIf image.GetPixel(x, y).G < 255 AndAlso image.GetPixel(x, y).R > 0 AndAlso image.GetPixel(x, y).R <= 255 Then
                                               map(x)(y) = New MapDot(False, New Point(x, y), False, True)
                                           Else
                                               map(x)(y) = New MapDot(False, New Point(x, y))
                                           End If
                                       End Sub)
            End Using
            Return dots
        End Function

        Public Sub SaveImage(map As MapDot()(), size As Size, savePath As String, Optional showWorking As Boolean = False)
            Using image As New Bitmap(size.Width, size.Height)
                Dim graphics As Graphics = Graphics.FromImage(image)
                Dim blackPen = New Pen(Brushes.Black)
                Dim yellowPen = New Pen(Brushes.Yellow)
                Dim whitePen = New Pen(Brushes.White)
                Dim redPen = New Pen(Brushes.Red)

                For x As Integer = 0 To size.Width - 1
                    For y As Integer = 0 To size.Height - 1
                        Dim mapDot As MapDot = map(x)(y)
                        Dim point = New Point(x, y)
                        Dim rectangle = New Rectangle(point, New Size(1, 1))
                        If mapDot.Wall Then
                            DrawPixel(graphics, blackPen, rectangle)
                        ElseIf mapDot.PathUsed Or mapDot.EndPoint Then
                            DrawPixel(graphics, redPen, rectangle)
                        ElseIf mapDot.EverBeenUsed AndAlso showWorking Then
                            DrawPixel(graphics, yellowPen, rectangle)
                        Else
                            DrawPixel(graphics, whitePen, rectangle)
                        End If
                    Next
                Next
                image.Save(savePath, ImageFormat.Png)
            End Using
        End Sub

        Private Sub DrawPixel(graphic As Graphics, pen As Pen, rectangle As Rectangle)
            graphic.FillRectangle(pen.Brush, rectangle)
            graphic.DrawRectangle(pen, rectangle)
        End Sub

    End Module
End Namespace

File: MazePicSolverCSharp/Helpers/ImageHelper.cs
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.Linq;

namespace MazePicSolverCSharp
{
    public class ImagerHelpers
    {
        public static MapDot[][] LoadMapDotsFromImage(string imagePath)
        {
            MapDot[][] dots;
            List<Color> r = new List<Color>();
            using (var image = new Bitmap(imagePath))
            {
                dots = Enumerable.Range(0, image.Size.Width).Select(t => new MapDot[image.Size.Height - 1]).ToArray();



                dots.IterateThroughMap((x, y, map) =>
                {
                    if (x == 14)
                    {
                        var b= true;
                    }
                    var pixel = image.GetPixel(x, y);
                    r.Add(pixel);
                    if (pixel.R == 0 && pixel.G == 0 && pixel.B == 0)
                    {
                        map[x][y] = new MapDot(true, new Point(x, y));
                    }
                    else if (pixel.R < 255 && pixel.G > 0 && pixel.G <= 255)
                    {
                        map[x][y] = new MapDot(false, new Point(x, y), true);
                    }
                    else if (pixel.G < 255 && pixel.R > 0 && pixel.R <= 255)
                    {
                        map[x][y] = new MapDot(false, new Point(x, y), false, true);
                    }
                    else
                    {
                        map[x][y] = new MapDot(false, new Point(x, y));
                    }
                });
            }

            var ber = r.Distinct().ToArray();
            return dots;
        }

        public static void SaveImage(MapDot[][] map, Size size, string savePath, bool showWorking = false)
        {
            using (var image = new Bitmap(size.Width, size.Height))
            {
                Graphics graphics = Graphics.FromImage(image);
                var blackPen = new Pen(Brushes.Black);
                var yellowPen = new Pen(Brushes.Yellow);
                var whitePen = new Pen(Brushes.White);
                var redPen = new Pen(Brushes.Red);
                for (int x = 0;
                     (x
                      <= (size.Width - 1));
                     x++)
                {
                    for (int y = 0;
                         (y
                          <= (size.Height - 1));
                         y++)
                    {
                        MapDot mapDot = map[x][y];
                        var point = new Point(x, y);
                        var rectangle = new Rectangle(point, new Size(1, 1));
                        if (mapDot.Wall)
                        {
                            DrawPixel(graphics, blackPen, rectangle);
                        }
                        else if ((mapDot.PathUsed || mapDot.EndPoint))
                        {
                            DrawPixel(graphics, redPen, rectangle);
                        }
                        else if ((mapDot.EverBeenUsed && showWorking))
                        {
                            DrawPixel(graphics, yellowPen, rectangle);
                        }
                        else
                        {
                            DrawPixel(graphics, whitePen, rectangle);
                        }
                    }
                }

                image.Save(savePath, ImageFormat.Png);
            }
        }

        private static void DrawPixel(Graphics graphic, Pen pen, Rectangle rectangle)
        {
            graphic.FillRectangle(pen.Brush, rectangle);
            graphic.DrawRectangle(pen, rectangle);
        }
    }
}